drivers/i2c/i2c_ll_stm32_v2.c

```diff
@@ -150,92 +150,80 @@ static int configure_dma(struct stream const *dma, struct dma_config *dma_cfg,
 
 	return 0;
 }
-#endif /* CONFIG_I2C_STM32_V2_DMA */
 
-static inline void msg_init(const struct device *dev, struct i2c_msg *msg,
-			    uint8_t *next_msg_flags, uint16_t slave,
-			    uint32_t transfer)
+static inline void dma_xfer_start(const struct device *dev, struct i2c_msg *msg)
 {
 	const struct i2c_stm32_config *cfg = dev->config;
 	struct i2c_stm32_data *data = dev->data;
 	I2C_TypeDef *i2c = cfg->i2c;
 
-	if (LL_I2C_IsEnabledReloadMode(i2c)) {
-		LL_I2C_SetTransferSize(i2c, msg->len);
+	if (msg->flags & I2C_MSG_READ) {
+		/* Configure RX DMA */
+		data->dma_blk_cfg.source_address = LL_I2C_DMA_GetRegAddr(
+			cfg->i2c, LL_I2C_DMA_REG_DATA_RECEIVE);
+		data->dma_blk_cfg.source_addr_adj = DMA_ADDR_ADJ_NO_CHANGE;
+		data->dma_blk_cfg.dest_address = (uint32_t)msg->buf;
+		data->dma_blk_cfg.dest_addr_adj = DMA_ADDR_ADJ_INCREMENT;
+		data->dma_blk_cfg.block_size = msg->len;
+
+		if (configure_dma(&cfg->rx_dma, &data->dma_rx_cfg,
+					&data->dma_blk_cfg) != 0) {
+			LOG_ERR("Problem setting up RX DMA");
```

Comment:
The error needs to be reported IMHO.
Maybe can be done in a later P-R, since this behavior was already there in original implementation.